### PR TITLE
sql: make the builtin hash functions variadic

### PR DESCRIPTION
**Pull request by danhhz**
Also sneak in sha512 while I'm here.

### CODE REVIEW

**Comment by danhhz** [comment]
The obvious counterproposal to the question I pose above is to leave everything as-is and add a parse_hex(string) -> bytes, which I am also willing to do if that's the consensus

File: pkg/sql/testdata/logic_test/builtin_function [hunk 1]
```diff
@@ -1519,3 +1519,40 @@ one
 one-three
 six-one-seven
 minus-two
+
+# Do some basic sanity checking of the variadic hash functions.
+query BBBB
+SELECT
+  sha512('1') = sha512('1'),
+  sha512('1') = sha512('2'),
+  sha512('1', '2') = sha512('1', '2'),
```
Comment (bdarnell) [inline pkg/sql/testdata/logic_test/builtin_function]: Should also test that `sha512('1', '2') = sha512('12')`

**Review by bdarnell — COMMENTED** [review]
All the hash functions should return the same type. In addition to the precedent set by postgresql's, `md5()`, we are now also bound by our own precedents in cockroach 1.0, so we're stuck with the hex encoding.

**Comment by danhhz** [comment]
Okay, works for me. Updated the commit message and PR description

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/testdata/logic_test/builtin_function, line 1528 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15828#-KjjhGFkMhw-zPK6ueQo-r1-1528:-KjjrDJ7Gg4-La-D3nh-:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/92fdb4d61645434e2754ad46949fe47f22650f2e/pkg/sql/testdata/logic_test/builtin_function#L1528)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Should also test that `sha512('1', '2') = sha512('12')`
</blockquote></details>

Done

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15828#-:-KjjrGzvxSGZSllrNVuA:b67gykp)*

**Comment by knz** [comment]
LGTM

But the real nice way to go a mile beyond pg here would be to provide both raw hash functions that return byte arrays, for performance, and keep the pg-named builtins that return hex-encoded strings alongside.

---

Reviewed 1 of 2 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15828#-:-KjmJGmAPHSbtHqbIFrT:b-d0fivp)*

**Comment by dt** [comment]
:lgtm:

If you felt like it, could sneak in crc and fnv impls while you're here since stdlib has hash.Hash impls for them too.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15828#-:-KjmViEmm62blbdGP9rB:bbsrdmu)*

**Comment by knz** [comment]
I'm not particularly fond of crc since it's so lousy, but people who want to work with legacy data may like it.

FNV on the other hand is a rather good hash, I second the proposal.

**Comment by petermattis** [comment]
> FNV on the other hand is a rather good hash, I second the proposal.

Every function we add extends the surface error we need to support. Just saying.

**Comment by danhhz** [comment]
Thanks for the reviews!

> Every function we add extends the surface area we need to support. Just saying.

=petermattis on this. I counterpropose that one of y'all open a PR with fnv if you'd like it for something. It's pretty easy to add new hash functions now

> But the real nice way to go a mile beyond pg here would be to provide both raw hash functions that return byte arrays, for performance, and keep the pg-named builtins that return hex-encoded strings alongside.h

Yeah, I may indeed do this. I'd really like the -> byte version myself for backup/restore testing, but I do think it's better for it to be a separate PR. Do you have a suggestion for the naming? `sha512_raw` / `sha512_bytes`?

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15828#-:-KjmmOw9yKpa2uoQ3LdZ:bg15jga)*

**Comment by knz** [comment]
_raw sounds best.

**Comment by danhhz** [comment]
👍  _raw is my inclination too

**Pull request merged**